[TECHDEBT][UTILITY] Fix Unit Tests using `testing.M`

Olshansk opened the pull request:
## Description

Fix unit tests that exist on the main branch.

## Origin Document

Fixes conversation in #259:

<img width="1059" alt="Screen Shot 2022-10-04 at 4 07 23 PM" src="https://user-images.githubusercontent.com/1892194/193946742-8a9f3914-86f1-4ccb-92fb-a1b025c5258a.png">

## Issue

Rather than the [core Go documentation](https://pkg.go.dev/testing), I believe [this article](https://medium.com/goingogo/why-use-testmain-for-testing-in-go-dafb52b406bc) does a better job at explaining why we had broken tests:

<img width="780" alt="Screen Shot 2022-10-04 at 4 12 38 PM" src="https://user-images.githubusercontent.com/1892194/193947366-5cd62ef3-a01a-4d74-95d3-45929ee9348a.png">

Before this change, even when a unit test failed, we still got a pass:

<img width="1241" alt="Screen Shot 2022-10-04 at 4 26 09 PM" src="https://user-images.githubusercontent.com/1892194/193948965-9079896e-5ab0-40dc-8f1b-f8bdcf4593c2.png">


## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [x] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

- Update `test_utility` target in the `Makefile`
- Not ignoring the exit code of `m.Run()` in all the files where we use it (several modules)
- Fix tests in the utility module:
    -  Remove redundant tests (e.g. `TestUtilityContext_UnstakesPausedBefore` which was replaced by `TestUtilityContext_UnstakePausedBefore`)
    - Do not ignore `m.Run` exit code so broken tests are caught
    - Improve readability of some tests (whitespace)
    - Do not unnecessarily export helpers
- Fix unit tests in the persistence module related to type casting

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [x] I have updated the corresponding README(s); local and/or global
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

Olshansk committed:
Fix utility test maketfile target

Olshansk committed:
Use the exit code from m.Run() everywhere where applicable

Olshansk committed:
Fix unit test

Olshansk committed:
Improve readability in some tests

Olshansk committed:
Remove TestUtilityContext_UnstakePausedBefore

Olshansk committed:
Remove redundant test

Olshansk committed:
Improve some test readability

Olshansk committed:
Update all the changelogs

Olshansk committed:
Merge branch 'main' into fix_all_main_tests

Olshansk committed:
Cleanup .json files correctly

andrewnguyen22 reviewed (changes requested):
The base change of catching the error code LGTM

However, you introduce a TON of white space into the tests. This makes me think we need a style guide or something because (a) it seems unnecessary to me (b) possibly out of scope?

I also see you marked the issue/PR as Utility, but you did this for all modules. Please update the issue

andrewnguyen22 commented on utility/test/actor_test.go:
```diff
@@ -215,21 +223,14 @@ func TestUtilityContext_BeginUnstakingMaxPaused(t *testing.T) {
 	}
 }
 
-func TestUtilityContext_CalculateRelays(t *testing.T) {
```
Why was this test removed?

Olshansk replied:
It was actually moved to a different part of the file, but I reverted it for clarity.

andrewnguyen22 commented on p2p/CHANGELOG.md:
```diff
@@ -7,8 +7,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [0.0.0.4] - 2022-10-04
```
OB1

Olshansk replied:
![OB1](https://user-images.githubusercontent.com/1892194/194115600-acc92a5c-317b-41dd-a80c-a0b94835cfc0.gif)

andrewnguyen22 reviewed (changes requested):
The base change of catching the error code LGTM

However, you introduce a TON of white space into the tests. This makes me think we need a style guide or something because (a) it seems unnecessary to me (b) possibly out of scope?

I also see you marked the issue/PR as Utility, but you did this for all modules. Please update the issue

Olshansk committed:
Merge with main

Olshansk committed:
Undo whitespace changes

Olshansk commented on utility/test/actor_test.go:
```diff
@@ -344,59 +353,32 @@ func TestUtilityContext_GetMessageEditStakeSignerCandidates(t *testing.T) {
 	}
 }
 
-func TestUtilityContext_UnstakesPausedBefore(t *testing.T) {
```
`TestUtilityContext_UnstakePausedBefore` covers this logic for all actor while `TestUtilityContext_UnstakesPausedBefore` only covers is for applications, and I believe is legacy code.

TestUtilityContext_UnstakesPausedBefore had a few issues:
1. Lack of casting on a line such as `require.Equal(t, -1, actor.GetUnstakingHeight())` (357)
2. The check on line 364 (`require.Equal(t, -1, actor.GetUnstakingHeight())`) was incorrect and simply had to be removed because on line 367 we do the right thing.

I fixed the tests initially to make it pass, but later realized that it's "legacy" relative to `TestUtilityContext_UnstakePausedBefore`.

Lmk if that makes sense.

Olshansk committed:
Update tests again

Olshansk commented:
@andrewnguyen22 I reverted whiteline changes.

> The base change of catching the error code LGTM

Ty.


> I also see you marked the issue/PR as Utility, but you did this for all modules. Please update the issue

Updated.

> However, you introduce a TON of white space into the tests. This makes me think we need a style guide or something because (a) it seems unnecessary to me (b) possibly out of scope?

Agreed that the whiteline changes are out of scope. Reverted.

Generally, I've been introducing whitelines between different "logical components" of the test. I agree that this is ill-defined and will update a best practice doc once I can quantity it.

E.g.

```go
			ctx := NewTestingUtilityContext(t, 0)

			actor := getFirstActor(t, ctx, actorType)
			addrBz, err := hex.DecodeString(actor.GetAddress())
			require.NoError(t, err)
			msgEditStake := &typesUtil.MessageEditStake{
				Address:   addrBz,
				Chains:    test_artifacts.DefaultChains,
				Amount:    test_artifacts.DefaultStakeAmountString,
				ActorType: actorType,
			}

			candidates, err := ctx.GetMessageEditStakeSignerCandidates(msgEditStake)
			require.NoError(t, err)
			require.Equal(t, len(candidates), 2, "unexpected number of candidates")
			require.Equal(t, actor.GetOutput(), hex.EncodeToString(candidates[0]), "incorrect output candidate")
			require.Equal(t, actor.GetAddress(), hex.EncodeToString(candidates[1]), "incorrect addr candidate")
			test_artifacts.CleanupTest(ctx)
```

```go
            // We get the context and used that to get the actor we use for the remainder of the test
			ctx := NewTestingUtilityContext(t, 0)
			actor := getFirstActor(t, ctx, actorType)

            // We get the address and verify there's no error
			addrBz, err := hex.DecodeString(actor.GetAddress())
			require.NoError(t, err)

            // We prepare a test artifact
			msgEditStake := &typesUtil.MessageEditStake{
				Address:   addrBz,
				Chains:    test_artifacts.DefaultChains,
				Amount:    test_artifacts.DefaultStakeAmountString,
				ActorType: actorType,
			}

            // We retrieve candidates, validate there's no error and run test checks using that return value
			candidates, err := ctx.GetMessageEditStakeSignerCandidates(msgEditStake)
			require.NoError(t, err)
			require.Equal(t, len(candidates), 2, "unexpected number of candidates")
			require.Equal(t, actor.GetOutput(), hex.EncodeToString(candidates[0]), "incorrect output candidate")
			require.Equal(t, actor.GetAddress(), hex.EncodeToString(candidates[1]), "incorrect addr candidate")

            // Test cleanup is a separate independent issue
			test_artifacts.CleanupTest(ctx)
```

Olshansk committed:
Merge with main

Olshansk committed:
Change the changelog dates

Olshansk committed:
Merge with main

Olshansk committed:
Merge with main and fix more unit tests that we introduced

Olshansk committed:
Fix another test

Olshansk committed:
Typo fix

andrewnguyen22 commented on persistence/shared_sql.go:
```diff
@@ -14,10 +14,10 @@ import (
-	UnstakingStatus        = 1
-	StakedStatus           = 2
-	UnstakedStatus         = 3
+	UndefinedStakingStatus = int32(0)
```
I'm not sure why this change was made within this scope and I don't see it reflected in the Changelog. Can you elaborate here please?

Olshansk replied:
The tests were breaking

Olshansk replied:
It can be reproduced via `go test -v -count=1 -v ./utility/...` on the main branch

Olshansk replied:
![Screen Shot 2022-10-06 at 12 01 35 PM](https://user-images.githubusercontent.com/1892194/194396958-123753be-c1d7-4f80-b3f7-a6ccbbff6e44.png)

Until we get this PR in, `make test_all` does not catch errors

andrewnguyen22 replied:
That's fine, but can you indicate in the changelog that you fixed broken unit tests? I'll go ahead and approve to not block

Olshansk replied:
Done

Olshansk committed:
Update changelogs

Olshansk committed:
Merge branch 'main' into fix_all_main_tests

Olshansk committed:
Update changelogs

Olshansk committed:
Typo
